### PR TITLE
Added clang-format

### PR DESCRIPTION
**Pull request by Legion2**
Added clang-format file with options to best match the current style of the project #385

### CODE REVIEW

**Comment by Legion2** [comment]
In the current clang-format config the opening braces are on the same line
```c++
if (foo) {

}
```
instead of next line
```c++
if (foo)
{

}
```

**Comment by Legion2** [comment]
Seems like the platformio compiler has problems with the `&` aligned to the right in the return type of functions such as `JsonObject &process_bledata(JsonObject &BLEdata) {`

```
/home/travis/build/1technophile/OpenMQTTGateway/main/ZgatewayBT.ino:641:53: error: 'process_bledata' was not declared in this scope
```

**Comment by Legion2** [comment]
@1technophile which pointer and reference alignment do you prefer:
- left `int* a;`
- middle `int * a;`
- right `int *a;`

**Comment by 1technophile** [comment]
> @1technophile which pointer and reference alignment do you prefer:

Take the one that you prefer, the 3 are ok for me.

**Comment by Legion2** [comment]
I don't understand why now this error occur:
```
/home/travis/build/1technophile/OpenMQTTGateway/main/main.ino:237:1: error: 'BLEdevice' does not name a type
 #endif
 ^
```

**Comment by Legion2** [comment]
platformio does not support .ino files correctly. According to https://github.com/platformio/platformio-core/issues/1683#issuecomment-640196593 we should use cpp files instead of the ino files.

**Comment by 1technophile** [comment]
> platformio does not support .ino files correctly. According to platformio/platformio-core#1683 (comment) we should use cpp files instead of the ino files.

Is it the add of clang that generated these issues? Before the building with PIO was working right.

**Comment by Legion2** [comment]
Yes it's because the magic behind the .ino files, which is handled correctly by Arduino IDE, is only a very bad workaround in PlatformIO . PlatformIO does not detect the functions definition in the ino files after reformatting, because the regex they use does not match on right aligned reference types. 
PlatformIO does not detect this as a function definition:
```c++
JsonObject &process_bledata(JsonObject &BLEdata)
{
...
```
So they don't apply the ino magic to the reformatted functions. It's a bug of PlatformIO, but they won't fix it.

**Comment by 1technophile** [comment]
Do you think that we could modify the clang definition so as to take into account the PIO "working" one as a bypass?

**Comment by Legion2** [comment]
I already modified the formatting so it works with PIO, but now there is another error I don't understand:
```
/home/travis/build/1technophile/OpenMQTTGateway/main/main.ino:237:1: error: 'BLEdevice' does not name a type
 #endif
 ^
```

`BLEdevice` is not used in main.ino

**Comment by 1technophile** [comment]
`BLEdevice is not used in main.ino`

From my previous experiments I think all the files are gathered into ones before building.

Can it be an issue with ZgatewayBT missing a new line at the end of the file?
Github gives a warning sign on this when you check the diff.

**Comment by 1technophile** [comment]
Of course it is not an EOL issue ..
.
If you modify this line:
`BLEdevice* getDeviceByMac(const char* mac) {`
to
`BLEdevice * getDeviceByMac(const char* mac) {`
or
`BLEdevice *getDeviceByMac(const char* mac) {`

It build...crazy isn't it?

on another topic I don't get why it build the BLE components for the esp32-RF gateway...

**Comment by Legion2** [comment]
I think this is also because of the .ino magic not corretly implemented by PIO.  The first version is detected by PIO as function declaration but then it does not find the type `BLEdevice` -> error. In the other two cases PIO don't detect it as functions, so it does not interpret it as type -> no error.

**Comment by 1technophile** [comment]
Hi @Legion2 Do you think you could finish the PR, I would like to push other modifications. And as this one is touching all the files, it would be better to avoid merge.

**Comment by Legion2** [comment]
@1technophile you can merge your changes, I can simply rebase this as all the changes are generated by clang-format, so this should not be a problem.

**Comment by 1technophile** [comment]
ok let's do it this way

**Comment by Legion2** [comment]
I have looked into the [InoToCPPConverter](https://github.com/platformio/platformio-core/blob/89cc6f9bf353bfe417f781de2a96b0dcf51674cf/platformio/builder/tools/piomisc.py#L32) of platformio and it's very fragile and has many bugs. It merges all .ino files, sorts them, so all files with something that looks like a `void setup` or `void loop` is considered the main ino file (the regex also matches commented out code!). Then they call the gcc preprocessor and then add function prototypes for all not declared functions. The prototypes are added at the beginning of the merged files without caring about declaration order. Also they ignore the declaration of structs and classes.

This causes `BLEdevice* getDeviceByMac(const char* mac);` to be declared before `BLEdevice` is included. Workaround add `BLEdevice* getDeviceByMac(const char* mac);` in `ZgatewayBT.ino`.

**Comment by 1technophile** [comment]
Wow, thanks for the analysis !

We would have to begin porting .ino to .cpp in the near future.

**Comment by Legion2** [comment]
I was able to analyse the temporarily generated `main.ino.cpp` file.

I will set the PointerAlignment to Left and then rebase the PR.

**Comment by Legion2** [comment]
I also add a .gitignore for the `.pio` and `.vscode` directories

**Comment by Legion2** [comment]
I use the [Clang-Format](https://marketplace.visualstudio.com/items?itemName=xaver.clang-format) Extension for VS Code.

File: main/main.ino [hunk 1]
```diff
@@ -29,129 +29,130 @@
...
-  #include <Preferences.h>
-  Preferences preferences;
-#ifdef MDNS_SD
-  #include <ESPmDNS.h>
-#endif
+#  include <ArduinoOTA.h>
+#  include <FS.h>
+#  include <WiFi.h>
+#  include <WiFiUdp.h>
+
```
[the remark this answers is not in the document]
Comment (Legion2) [reply main/main.ino:132]: I thinks this is because of the sorting of the different includes

[the remark this answers is not in the document]
Comment (Legion2) [reply main/main.ino:132]: but of course we can change this, there are many options to format the includes

[the remark this answers is not in the document]
Comment (1technophile) [reply main/main.ino:132]: ok we let it like that

**Comment by 1technophile** [comment]
Should we not add clang in the recommended extension list of vscode + auto formating on document save?

**Comment by Legion2** [comment]
It would make sense to update the "Development contributions" documentation and add how to do the formatting. I personally don't use the auto formatting on save. But at least we should document how to enable it. @1technophile What is your development setup? Is there already a vscode section for documentation? 

I will also look into converting the travis ci into GitHub actions, so there is a consistent experience and maybe split the tests into parallel jobs so this don't take 20 min.

**Comment by 1technophile** [comment]
> It would make sense to update the "Development contributions" documentation and add how to do the formatting. 

I was thinking on the same thing yesterday ;-)

For the moment we have:
https://docs.openmqttgateway.com/participate/development.html

But no special sections for VSCode (only the [upload ](https://docs.openmqttgateway.com/upload/pio.html)doc)

**Comment by 1technophile** [comment]
>  I will also look into converting the travis ci into GitHub actions, so there is a consistent experience and maybe split the tests into parallel jobs so this don't take 20 min.

It would be nice !

**Pull request merged**